src/main/java/com/esri/core/geometry/ogc/OGCConcreteGeometryCollection.java

```diff
@@ -501,4 +519,303 @@ public int hashCode() {
+					break;
+				}
+			}
+			
+			if (!good)
+				return false;
+		}
+		
+		//each geometry of another is contained in a geometry from this.
+		return true;
+	}
+	
+	/**
+	 * Checks if collection is flattened.
+	 * @return True for the flattened collection. A flattened collection contains up to three non-empty geometries:
+	 * an OGCMultiPoint, an OGCMultiPolygon, and an OGCMultiLineString.
+	 */
+	public boolean isFlattened() {
+		int n = numGeometries();
+		if (n > 3)
+			return false;
+
+		for (int i = 0; i < n; ++i) {
+			OGCGeometry g = geometryN(i);
+			if (g.isEmpty())
+				return false;//no empty allowed
+			
+			String t = g.geometryType();
+			if (t != OGCMultiPoint.TYPE && t != OGCMultiPolygon.TYPE && t != OGCMultiLineString.TYPE)
+				return false;
+		}
+		
+		return true;
+	}
+
+	/**
+	 * Flattens Geometry Collection.
+	 * The result collection contains up to three geometries:
+	 * an OGCMultiPoint, an OGCMultiPolygon, and an OGCMultilineString.
+	 * @return A flattened Geometry Collection, or self if already flattened.
+	 */
+	public OGCConcreteGeometryCollection flatten() {
+		if (isFlattened()) {
+			return this;
+		}
+		
+		OGCMultiPoint multiPoint = null;
+		ArrayList<Geometry> polygons = null;
+		OGCMultiLineString polyline = null;
+		GeometryCursor gc = getEsriGeometryCursor();
+		for (Geometry g = gc.next(); g != null; g = gc.next()) {
+			if (g.isEmpty())
+				continue;
+			
+			Geometry.Type t = g.getType();
+
+			if (t == Geometry.Type.Point) {
+				if (multiPoint == null) {
+					multiPoint = new OGCMultiPoint(esriSR);
+				}
+
+				((MultiPoint)multiPoint.getEsriGeometry()).add((Point)g);
+				continue;
+			}
+
+			if (t == Geometry.Type.MultiPoint) {
+				if (multiPoint == null)
+					multiPoint = new OGCMultiPoint(esriSR);
+				
+				((MultiPoint)multiPoint.getEsriGeometry()).add((MultiPoint)g, 0, -1);
+				continue;
+			}
+			
+			if (t == Geometry.Type.Polyline) {
+				if (polyline == null)
+					polyline = new OGCMultiLineString(esriSR);
+				
+				((MultiPath)polyline.getEsriGeometry()).add((Polyline)g, false);
+				continue;
+			}
+			
+			if (t == Geometry.Type.Polygon) {
+				if (polygons == null)
+					polygons = new ArrayList<Geometry>();
+				
+				polygons.add(g);
+				continue;
+			}
+			
+			throw new GeometryException("internal error");//what else?
+		}
+
+		List<OGCGeometry> list = new ArrayList<OGCGeometry>();
+
+		if (multiPoint != null)
+			list.add(multiPoint);
+
+		if (polyline != null)
+			list.add(polyline);
+		
+		if (polygons != null) {
+			GeometryCursor unionedPolygons = OperatorUnion.local().execute(new SimpleGeometryCursor(polygons), esriSR, null);
+			Geometry g = unionedPolygons.next();
+			if (!g.isEmpty()) {
+				list.add(new OGCMultiPolygon((Polygon)g, esriSR));
+			}
+
+		}
+
+		return new OGCConcreteGeometryCollection(list, esriSR);
+	}
+	
+	/**
+	 * Fixes topological overlaps in the GeometryCollecion.
+	 * This is equivalent to union of the geometry collection elements.
```

Comment:
`flatten` method already performs union on polygons; should `flatten` also union points and line strings?